core/src/main/java/org/apache/calcite/rel/rules/AggregateRemoveRule.java

```diff
@@ -58,37 +68,63 @@ public AggregateRemoveRule(Class<? extends Aggregate> aggregateClass,
     // distinct to make it correct up-front, we can get rid of the reference
     // to the child here.
     super(
-        operand(aggregateClass,
+        operandJ(aggregateClass, null, agg -> isAggregateSupported(agg),
             operand(RelNode.class, any())),
         relBuilderFactory, null);
   }
 
+  private static boolean isAggregateSupported(Aggregate aggregate) {
+    if (aggregate.indicator
+        || aggregate.getGroupType() != Aggregate.Group.SIMPLE) {
+      return false;
+    }
+    // If any aggregate functions do not support splitting, bail out
+    for (AggregateCall aggregateCall : aggregate.getAggCallList()) {
+      if (aggregateCall.filterArg >= 0
+          || aggregateCall.getAggregation()
+              .unwrap(SqlSplittableAggFunction.class) == null) {
+        return false;
+      }
+    }
+    return true;
+  }
+
   //~ Methods ----------------------------------------------------------------
 
   public void onMatch(RelOptRuleCall call) {
     final Aggregate aggregate = call.rel(0);
     final RelNode input = call.rel(1);
-    if (!aggregate.getAggCallList().isEmpty() || aggregate.indicator) {
-      return;
-    }
     final RelMetadataQuery mq = call.getMetadataQuery();
     if (!SqlFunctions.isTrue(mq.areColumnsUnique(input, aggregate.getGroupSet()))) {
       return;
     }
-    // Distinct is "GROUP BY c1, c2" (where c1, c2 are a set of columns on
-    // which the input is unique, i.e. contain a key) and has no aggregate
-    // functions. It can be removed.
-    final RelNode newInput = convert(input, aggregate.getTraitSet().simplify());
 
-    // If aggregate was projecting a subset of columns, add a project for the
-    // same effect.
+    final RexBuilder rexBuilder = aggregate.getCluster().getRexBuilder();
+    final List<RexNode> projects = new LinkedList<>();
+    for (AggregateCall aggCall : aggregate.getAggCallList()) {
```

Comment:
it's unusual to use LinkedList; since you're only appending, ArrayList would be the least surprising choice